### PR TITLE
Global bucket

### PR DESCRIPTION
**Pull request by dariajung**
Started playing around with the global bucket renaming, and using the TypeChecker to get information on the symbol to see if renaming is possible. 

I think the` if` statements are a bit messy. My intent in that was I was trying to avoid `case` statements to keep text splicing to a minimum.

Doc to go with PR: https://docs.google.com/document/d/1qmBQfTcJMLv3OM4YeyIWSEb79dd8SQWNf2TF6PHJJNE/edit

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 import * as ts from 'typescript';
 
-const DEBUG = false;
+const DEBUG = true;
```
Comment (mprobst) [inline src/main.ts]: Remember to toggle off before pushing.

File: src/main.ts [hunk 2]
```diff
@@ -12,6 +12,8 @@ export const options: ts.CompilerOptions = {
 
 export class Minifier {
   static reservedJSKeywords = Minifier.buildReservedKeywordsMap();
+  private renameMap: {[name: string]: string} = {};
```
Comment (mprobst) [inline src/main.ts]: Could you add comments as to what exactly the keys and values are (e.g. theoretically the keys could be fully qualified names + properties or something like that)?

File: src/main.ts [hunk 3]
```diff
@@ -80,6 +115,12 @@ export class Minifier {
     }
   }
 
+  private renameIdent(node: ts.Node) {
+    return this.renameProperty(node.getText());
```
Comment (mprobst) [inline src/main.ts]: should this call this.ident(node) inside, in case you change that at some point?

File: src/main.ts [hunk 4]
```diff
@@ -37,16 +39,49 @@ export class Minifier {
...
-        output += propAccessExp.name.text + '$mangled';
-        return output;
+      case ts.SyntaxKind.Identifier: {
+        let parent = node.parent;
+
+        if (parent.kind === ts.SyntaxKind.PropertyDeclaration) {
+          return this.renameIdent(node);
+        } else if (parent.kind === ts.SyntaxKind.PropertyAccessExpression) {
+          let pae = <ts.PropertyAccessExpression>parent;
+          let exprSymbol = typeChecker.getTypeAtLocation(pae.expression).symbol;
```
Comment (mprobst) [inline src/main.ts]: Extract a method for getExprSymbol, and add comments why there are two attempts?

File: src/main.ts [hunk 5]
```diff
@@ -37,16 +39,49 @@ export class Minifier {
 
   // Recursively visits every child node, emitting text of the sourcefile that is not a part of
   // a child node.
-  visit(node: ts.Node) {
+  visit(node: ts.Node, typeChecker?: ts.TypeChecker) {
```
Comment (mprobst) [inline src/main.ts]: Why do you pass down the typeChecker? Couldn't that be a field on the class? It doesn't change during the recursion, does it?

File: src/main.ts [hunk 6]
```diff
@@ -37,16 +39,49 @@ export class Minifier {
...
-        output += this.visit(propAccessExp.expression);
-        output += '.';
-        // Adds '$mangled' to an identifier to ensure that the identifier is being altered.
-        output += propAccessExp.name.text + '$mangled';
-        return output;
+      case ts.SyntaxKind.Identifier: {
+        let parent = node.parent;
+
+        if (parent.kind === ts.SyntaxKind.PropertyDeclaration) {
+          return this.renameIdent(node);
```
Comment (mprobst) [inline src/main.ts]: Needs a comment - why can we just rename property declarations here without the type check below? (I know now, but I won't in 6 months...).

File: src/main.ts [hunk 7]
```diff
@@ -37,16 +39,49 @@ export class Minifier {
...
+          let pae = <ts.PropertyAccessExpression>parent;
+          let exprSymbol = typeChecker.getTypeAtLocation(pae.expression).symbol;
+          if (!exprSymbol) {
+            exprSymbol = typeChecker.getSymbolAtLocation(pae.expression);
+          }
+          if (!exprSymbol) {
+            exprSymbol = typeChecker.getSymbolAtLocation(pae);
+          }
+          let childText = '';
+          if (exprSymbol) {
```
Comment (mprobst) [inline src/main.ts]: You can make your code easier to read and less indented by doing early exits for error cases.

E.g. here say `if (!exprSymbol) return this.renameIdent(node);`, and then continue with the code underneath.

File: src/main.ts [hunk 8]
```diff
@@ -37,16 +39,49 @@ export class Minifier {
...
+            // start off by assuming the property is rename-able
+            let rename: boolean = true;
+
+            // check if a source filename of a declaration ends in .d.ts
+            exprSymbol.declarations.forEach((decl) => {
```
Comment (mprobst) [inline src/main.ts]: if you rephrase this as a `for (var decl of exprSymbol.declarations)` loop, you could use the same technique as mentioned above for an early return.

File: src/main.ts [hunk 9]
```diff
@@ -37,16 +39,49 @@ export class Minifier {
...
+              }
+            });
+
+            if (rename) {
+              childText = this.renameIdent(node);
+            } else {
+              childText = this.ident(node);
+            }
+          } else {
+            childText = this.renameIdent(node);
```
Comment (mprobst) [inline src/main.ts]: Isn't this branch dangerous somehow? We found a property use site, but we have no idea where it's coming from, could be external or internal? If we guess incorrectly, we could break the program. Should this rather report an error?

File: src/main.ts [hunk 10]
```diff
@@ -37,16 +39,49 @@ export class Minifier {
...
     switch (node.kind) {
-      case ts.SyntaxKind.PropertyAccessExpression: {
-        let output = '';
-        let propAccessExp = <ts.PropertyAccessExpression>node;
-        output += this.visit(propAccessExp.expression);
-        output += '.';
-        // Adds '$mangled' to an identifier to ensure that the identifier is being altered.
-        output += propAccessExp.name.text + '$mangled';
-        return output;
+      case ts.SyntaxKind.Identifier: {
```
Comment (mprobst) [inline src/main.ts]: I don't understand why you're no longer casing for the PropertyAccessExpression - the code within this case block is nearly all about the PropertAccessExpression, and the other kinds of parents are all special cased below again. Wouldn't this be simpler if you just added them as separate cases?

File: src/main.ts [hunk 11]
```diff
@@ -121,6 +162,15 @@ export class Minifier {
     return Minifier.reservedJSKeywords.hasOwnProperty(str);
   }
 
+  renameProperty(name: string): string {
+    if (!this.renameMap.hasOwnProperty(name)) {
+      let newName = this.generateNextPropertyName(this.lastGeneratedPropName);
```
Comment (mprobst) [inline src/main.ts]: would it be less error prone to do the assignment into `this.lastGeneratedPropName` within the method `generateNextPropertyName`? Then the method would be in control of its invariant, and it would be more obvious how it works.

File: test/unit/main_test.ts [hunk 12]
```diff
@@ -64,13 +75,30 @@ describe('Recognizes invalid TypeScript inputs', () => {
...
+        .to.equal('for (var x in $._) { var y = $._.a; }');
+    expectTranslate('class Foo {bar: string;} class Baz {bar: string;}')
+        .to.equal('class Foo {$: string;} class Baz {$: string;}');
+  });
+});
+
+describe('Selective renaming', () => {
+  it('does not rename property names from the standard library', () => {
+    expectTranslate('Math.random();').to.equal('Math.random();');
+    expectTranslate('console.log();').to.equal('console.log();');
```
Comment (mprobst) [inline test/unit/main_test.ts]: I think one or two test cases (static method `Math.random` plus instance method `"foo".substring`, maybe plus a DOM method) would suffice here. Good testing is about testing different cases and situations, not just having a lot of test cases that essentially test the same.

Think about this as trying to find tests that cover each line of your code, each branch of an if statement, and (as much as possible) each permutation of the branches. You should have test cases to demonstrate each interesting branch works, but ideally only have exactly one test case - all other tests are just waste, as they do the same thing over and over again.

File: src/main.ts (hunk 1 above)
Comment (dariajung) [inline src/main.ts]: changed back to `false`

File: src/main.ts (hunk 3 above)
Comment (dariajung) [inline src/main.ts]: changed to `return this.renameProperty(this.ident(node));`

File: src/main.ts (hunk 5 above)
Comment (dariajung) [inline src/main.ts]: Added typeChecker field on `Minifier` and added `setTypeChecker()` method on `Minifier` class.

File: src/main.ts (hunk 2 above)
Comment (dariajung) [inline src/main.ts]: Added.

File: src/main.ts (hunk 10 above)
Comment (dariajung) [inline src/main.ts]: Talked offline about this: I wanted to avoid having to do string splicing. 

Possible solution is to create a method that does what the code that's currently under the `default` case does, and move these `if`s to `case` statements.

File: src/main.ts (hunk 7 above)
Comment (dariajung) [inline src/main.ts]: Done.

File: src/main.ts (hunk 8 above)
Comment (dariajung) [inline src/main.ts]: Done.

File: src/main.ts (hunk 9 above)
Comment (dariajung) [inline src/main.ts]: Throws error: `throw new Error('Symbol information could not be extracted.\n');`

File: src/main.ts (hunk 11 above)
Comment (dariajung) [inline src/main.ts]: Done (though this adds three places before `return` statements where assignment needs to be done)

File: test/unit/main_test.ts (hunk 12 above)
Comment (dariajung) [inline test/unit/main_test.ts]: Removed extraneous tests.

File: src/main.ts (hunk 4 above)
Comment (dariajung) [inline src/main.ts]: Added `getExprSymbol` method

File: src/main.ts [hunk 13]
```diff
@@ -35,17 +40,63 @@ export class Minifier {
...
-        output += propAccessExp.name.text + '$mangled';
+        let children = pae.getChildren();
+
+        for (var child of children) {
+          if (child.kind !== ts.SyntaxKind.Identifier) {
+            output += this.visit(child);
+          } else {
+            // Early exit when exprSymbol is undefined.
+            if (!exprSymbol) {
+              throw new Error('Symbol information could not be extracted.\n');
```
Comment (mprobst) [inline src/main.ts]: move this to a reportError(n: ts.Node, msg: message) style function?

File: src/main.ts [hunk 14]
```diff
@@ -35,17 +40,63 @@ export class Minifier {
...
+        let exprSymbol = this.getExpressionSymbol(pae);
         let output = '';
-        let propAccessExp = <ts.PropertyAccessExpression>node;
-        output += this.visit(propAccessExp.expression);
-        output += '.';
-        // Adds '$mangled' to an identifier to ensure that the identifier is being altered.
-        output += propAccessExp.name.text + '$mangled';
+        let children = pae.getChildren();
+
+        for (var child of children) {
```
Comment (mprobst) [inline src/main.ts]: Do you need the loop here? You know exactly what children a PAE has, don't you? What about just appending the text up to `pae.identifer.getStart()`, then doing the renaming, and be done?

File: src/main.ts [hunk 15]
```diff
@@ -80,6 +131,26 @@ export class Minifier {
     }
   }
 
+  private getExpressionSymbol(node: ts.PropertyAccessExpression) {
+    // gets information about symbol
+    let exprSymbol = this.typeChecker.getSymbolAtLocation(node.expression); 
+
+    // USEFUL LATER: to get symbol information of a type
```
Comment (mprobst) [inline src/main.ts]: Don't check in code that might just be useful later, it'll clutter up your code base.

File: src/main.ts [hunk 16]
```diff
@@ -80,6 +131,26 @@ export class Minifier {
     }
   }
 
+  private getExpressionSymbol(node: ts.PropertyAccessExpression) {
+    // gets information about symbol
```
Comment (mprobst) [inline src/main.ts]: Drop the comment?

File: src/main.ts [hunk 17]
```diff
@@ -80,6 +131,26 @@ export class Minifier {
...
+    let exprSymbol = this.typeChecker.getSymbolAtLocation(node.expression);
```
Comment (mprobst) [inline src/main.ts]: Don't you want the symbol for the property, i.e. the RHS of the expression?

File: src/main.ts (hunk 14 above)
Comment (dariajung) [inline src/main.ts]: Fixed.

File: src/main.ts (hunk 15 above)
Comment (dariajung) [inline src/main.ts]: Removed

File: src/main.ts (hunk 16 above)
Comment (dariajung) [inline src/main.ts]: Dropped.

File: src/main.ts (hunk 17 above)
Comment (dariajung) [inline src/main.ts]: Changed to `node.name`.

File: src/main.ts [hunk 18]
```diff
@@ -35,17 +41,69 @@ export class Minifier {
...
+  setTypeChecker(typeChecker: ts.TypeChecker) { this.typeChecker = typeChecker; }
+
+  reportError(n: ts.Node, message: string) {
+    var file = n.getSourceFile();
+    var fileName = file.fileName;
+    var start = n.getStart(file);
+    var pos = file.getLineAndCharacterOfPosition(start);
+    var fullMessage = `${fileName}:${pos.line + 1}:${pos.character + 1}: ${message}`;
+    throw new Error(fullMessage);
+    this.errors.push(fullMessage);
```
Comment (mprobst) [inline src/main.ts]: this line can never be reached, can it?

File: src/main.ts [hunk 19]
```diff
@@ -35,17 +41,69 @@ export class Minifier {
...
+
+        // Early exit when exprSymbol is undefined.
+        if (!exprSymbol) {
+          this.reportError(pae.name, 'Symbol information could not be extracted.\n');
+        } else {
+          // start off by assuming the property is rename-able
+          let rename: boolean = true;
+
+          // check if a source filename of a declaration ends in .d.ts
+          for (var decl of exprSymbol.declarations) {
```
Comment (mprobst) [inline src/main.ts]: what about:

```ts
var isExternal = exprSymbol.declarations.some((d) => d.getSourceFile().fileName.match(/\.d\.ts$/);
if (isExternal) return output + this.ident(pae.name);
return output + this.renameIdent(...);
```

File: src/main.ts [hunk 20]
```diff
@@ -35,17 +41,69 @@ export class Minifier {
...
+        let exprSymbol = this.getExpressionSymbol(pae);
+        let output = '';
+        let children = pae.getChildren();
+
+        output += this.visit(pae.expression);
+        output += pae.dotToken.getText();
+
+        // Early exit when exprSymbol is undefined.
+        if (!exprSymbol) {
+          this.reportError(pae.name, 'Symbol information could not be extracted.\n');
```
Comment (mprobst) [inline src/main.ts]: just return here and avoid indenting further?

File: src/main.ts (hunk 19 above)
Comment (dariajung) [inline src/main.ts]: Whoa, I didn't know JS had `some`. That's nifty.

File: src/main.ts (hunk 20 above)
Comment (dariajung) [inline src/main.ts]: `return;` added

File: src/main.ts [hunk 21]
```diff
@@ -35,17 +45,58 @@ export class Minifier {
...
+          return;
+        }
+
+        var isExternal = exprSymbol.declarations.some(
+            (decl) => !!(decl.getSourceFile().fileName.match(/\.d\.ts/)));
+        if (isExternal) {
```
Comment (mprobst) [inline src/main.ts]: nit: prefer to keep if statements that fit on one line on one line (for single statement, no braces, no else ifs).

**Comment by mprobst** [comment]
LGTM.
